Parser/Token.cs

```diff
@@ -38,9 +39,17 @@ public enum E {
    // Utility function used to echo an error to the console
    public void PrintError () {
       if (Kind != ERROR) throw new Exception ("PrintError called on a non-error token");
+      int start = Line > 3 ? Line - 3 : 0, end = Line + 2 <= Source.Lines.Length ? Line + 2 : Source.Lines.Length, line = start + 1;
+      Console.WriteLine ($"File: {Source.FileName}\n───┬────────────────");
+      foreach (var str in Source.Lines[start..Line]) Console.WriteLine ($"{line++,3}│ {str}");
       Console.ForegroundColor = ConsoleColor.Yellow;
-      Console.WriteLine ($"At line {Line}, column {Column} of {Source.FileName}: {Text}");
+      Console.WriteLine ("^", Console.CursorLeft = Column + 4);
+      Console.WriteLine ($"{ErrorMessage}", Console.CursorLeft = Math.Max (0, Column - ErrorMessage.Length / 2) + 5);
       Console.ResetColor ();
+      foreach (var str in Source.Lines[Line..end]) {
+         if (str is "\r") break;
+         Console.WriteLine ($"{line++,3}│ {str}");
+      }

```

Comment:
Here there are 2 for loops; one to render top 2 lines and the other to render bottom 2 lines. Can this be done in a single loop?